### PR TITLE
upload folder to workspace

### PR DESCRIPTION
**Pull request by Aniket-Engg**
related to #413 

1. Allows uploading a folder into current workspace
2. Fixes the tooltip message for file explorer icons
3. Fixes the overwrite modal not appearing for all uploaded files
4. Adds matomo tracking for file and folder upload
5. Refactoring

### CODE REVIEW

File: apps/remix-ide/src/app/tabs/locales/en/filePanel.json [hunk 1]
```diff
@@ -39,10 +39,11 @@
...
   "filePanel.compileForNahmii": "Compile for Nahmii",
-  "filePanel.createNewFile": "Create New File",
-  "filePanel.createNewFolder": "Create New Folder",
-  "filePanel.publishToGist": "Publish all the current workspace files to a github gist",
-  "filePanel.uploadFile": "Load a local file into current workspace",
+  "filePanel.createNewFile": "Create new file",
+  "filePanel.createNewFolder": "Create new folder",
+  "filePanel.publishToGist": "Publish current workspace to GitHub gist",
+  "filePanel.uploadFile": "Upload files into current workspace",
+  "filePanel.uploadFolder": "Upload folder into current workspace",
```
Comment (LianaHus) [inline apps/remix-ide/src/app/tabs/locales/en/filePanel.json:46]: There is an inconsistency here. I don't think we need "current workspace" here.

File: libs/remix-ui/workspace/src/lib/actions/workspace.ts [hunk 2]
```diff
@@ -358,46 +359,83 @@ export const switchToWorkspace = async (name: string) => {
...
-        if ((config.get('currentFile') === name) && (editor.currentContent() !== event.target.result)) {
-          editor.setText(name, event.target.result)
-        }
+    if (!await workspaceProvider.exists(name)) {
+      loadFile(name, file, workspaceProvider, cb)
+    } else {
+      const modalContent: AppModal = {
+        id: 'overwriteUploadFile',
+        title: 'Confirm overwrite',
+        message: `The file ${name} already exists! Would you like to overwrite it?`,
```
Comment (LianaHus) [inline libs/remix-ui/workspace/src/lib/actions/workspace.ts:401]: - First of all the file name should be in quotes.

- I see a problem with OK/Cancel choices. Why not add "(Number)" and allow to upload. We should offer a solution to the user.

File: libs/remix-ui/workspace/src/lib/components/file-explorer-menu.tsx [hunk 3]
```diff
@@ -10,27 +10,33 @@ export const FileExplorerMenu = (props: FileExplorerMenuProps) => {
     menuItems: [
       {
         action: 'createNewFile',
-        title: 'Create New File',
+        title: 'Create new file',
```
Comment (LianaHus) [inline libs/remix-ui/workspace/src/lib/components/file-explorer-menu.tsx:13]: I think you should use names from apps/remix-ide/src/app/tabs/locales/en/filePanel.json

**Review by LianaHus — DISMISSED** [review]

File: apps/remix-ide/src/app/tabs/locales/en/filePanel.json (hunk 1 above)
Comment (Aniket-Engg) [reply apps/remix-ide/src/app/tabs/locales/en/filePanel.json:46]: I agree

File: libs/remix-ui/workspace/src/lib/actions/workspace.ts (hunk 2 above)
Comment (Aniket-Engg) [reply libs/remix-ui/workspace/src/lib/actions/workspace.ts:401]: First is fine. Second is an existing behaviour. I just refactored this code.

File: libs/remix-ui/workspace/src/lib/actions/workspace.ts [hunk 4]
```diff
@@ -408,9 +420,20 @@ export const uploadFolder = async (target, targetFolder: string, cb?: (err: Erro
...
+        modalType: ModalTypes.confirm,
+        okLabel: 'OK',
+        cancelLabel: 'Cancel',
+        okFn: () => {
+          loadFile(name, file, workspaceProvider, cb)
+        },
+        cancelFn: () => {},
+        hideFn: () => {}
+      }
+      plugin.call('notification', 'modal', modalContent)
```
[the remark this answers is not in the document]
Comment (Aniket-Engg) [reply libs/remix-ui/workspace/src/lib/actions/workspace.ts:436]: For now, it is showing modal for each overwriting file one by one. What do you suggest to improve?

File: libs/remix-ui/workspace/src/lib/components/file-explorer-menu.tsx (hunk 3 above)
Comment (Aniket-Engg) [reply libs/remix-ui/workspace/src/lib/components/file-explorer-menu.tsx:13]: ok

File: libs/remix-ui/workspace/src/lib/actions/workspace.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (yann300) [reply libs/remix-ui/workspace/src/lib/actions/workspace.ts:436]: if it's working like that, it's ok

File: libs/remix-ui/workspace/src/lib/components/file-explorer-menu.tsx (hunk 3 above)
Comment (Aniket-Engg) [reply libs/remix-ui/workspace/src/lib/components/file-explorer-menu.tsx:13]: actually that is being used for `defaultMessage` , tooltip text is coming for fielPanel.json:
`tooltipText={<FormattedMessage id={filePanel.${action}} defaultMessage={title} />}`

File: apps/remix-ide/src/app/tabs/locales/en/filePanel.json (hunk 1 above)
Comment (Aniket-Engg) [reply apps/remix-ide/src/app/tabs/locales/en/filePanel.json]: updated

File: libs/remix-ui/workspace/src/lib/actions/workspace.ts (hunk 2 above)
Comment (Aniket-Engg) [reply libs/remix-ui/workspace/src/lib/actions/workspace.ts]: first is addressed. Second can come in a different PR

Comment (Aniket-Engg) [reply libs/remix-ui/workspace/src/lib/actions/workspace.ts]: created issued for second one: https://github.com/ethereum/remix-project/issues/3456

**Review by yann300 — APPROVED** [review]

**Pull request merged**